### PR TITLE
+str Add onErrorComplete operator.

### PR DESCRIPTION
**Pull request by He-Pin**
https://github.com/akka/akka/issues/31671

@marcospereira @ncreep Would you like to give some inputs?

Have to duplicate some code, as the ambiguous reference to overloaded definition issue.

Still need to add some tests.

### CODE REVIEW

**Comment by ncreep** [comment]
Hi @He-Pin,

Thanks for the effort, but this feature doesn't really address my original request. One of my requirements is that I didn't want to catch exceptions blindly, but rather avoid throwing them in the first place (as `idle-timeout`) does.

**Comment by He-Pin** [comment]
I think we can add a akka's timeout exception,which extends the current TimeoutExpection？

**Comment by He-Pin** [comment]
@ncreep I want to compose https://github.com/akka/akka/pull/31640 with this pr, what do you think about it?

**Comment by ncreep** [comment]
#31640 does resolve my issue with the exception (if you add the `recoverWithComplete` close enough to the `idle` component). But I'd rather avoid using exceptions for flow control. Seeing how in my use-case having the timeout is not at all "exceptional", I think it would be nicer to have a non-exceptional way to create completion.

Having said, I think the `recoverWithComplete` is generally useful, and I don't have any other objections.

**Comment by He-Pin** [comment]
@ncreep Yes, I think the current idleTimeout can accept an alternative Source when idle. if the alternative source is not provided, then just fail the source.
then `idleTimeout(timeout, Source.empty)` will address your origin issue.

**Comment by He-Pin** [comment]
@ncreep Yes, I think the current idleTimeout can accept an alternative Source when idle. if the alternative source is not provided, then just fail the source.
then `idleTimeout(timeout, Source.empty)` will address your origin issue.

**Comment by ncreep** [comment]
@He-Pin, that does sound good. But I wonder whether there's something more general that can be done here. Some implementation that won't be specific to just the `idleTimeout` use-case. I guess that thinking about this is what got the original issue side-tracked...

**Review by johanandren — COMMENTED** [review]
I like this, together with #31640 it nicely covers the various timeout scenarios. 

I wonder if we should skip the no-param variation though, handing it `Throwable` in user code seems concise enough for the catch-all IMO.

Also not sure about the name, should it rather be something like `completeOn[IdleTimeoutException]`? Recover would be to turn the stream back into something working, but this operator always completes.

**Comment by He-Pin** [comment]
Cool, I was following the old name, your `completeOn` is definitely a better name, the name in reactor-core is `onErrorComplete`.

**Comment by He-Pin** [comment]
@ncreep I was thinking about that too, but never came out a better idea. Will try to get this merged first:) do you have something in mind?

**Comment by He-Pin** [comment]
Still on holiday, will find more time to get it ready.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+          case _    => failStage(ex)
+        } catch {
+          case NonFatal(ex) => failStage(ex)
+        }
+
+      override def onPull(): Unit = pull(in)
+
+      setHandlers(in, out, this)
+    }
+}
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala:40]: I think we should do with existing operators `recoverWith { ... => Stream.empty }` instead of a custom operator for this.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -2151,8 +2151,12 @@ private[akka] object TakeWithin {
...
       def onFailure(ex: Throwable): Unit =
         if ((maximumRetries < 0 || attempt < maximumRetries) && pf.isDefinedAt(ex)) {
-          switchTo(pf(ex))
-          attempt += 1
+          pf(ex) match {
+            case source if source eq Source.empty => completeStage()
+            case other =>
+              switchTo(other)
+              attempt += 1
+          }
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2159]: Looks like an unrelated change?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala (hunk 1 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala:40]: Yes, that's my previous version, but which will need a `pf.andThen` seems a little complicate too.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 2 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2159]: When the fallback source is empty then avoid the materialization?

**Comment by ncreep** [comment]
> @ncreep I was thinking about that too, but never came out a better idea. Will try to get this merged first:) do you have something in mind?

@He-Pin, no, unfortunately I'm not proficient enough in the underlying mechanics of Akka streams to have something concrete in mind. 

But I can imagine that whatever the general solution would be it would have to include some change to the underlying signalling mechanism. So that you can "catch" completion signals and then convert those into exceptions. So on its own, `idleTimeout` would send a completion signal, and then if you want the current behavior back you wrap that signal and throw a `Timeout` exception for it. That would require some way to identify the origin of the completion signal, so that you don't accidentally catch signals that you didn't mean to. Though maybe something more local would be better, so that every component would receive an `onComplete` handler, which can decide what to do with the completion of the underlying component (all this assumes that the direction of `complete -> exception` is more "natural" than `exception -> complete`).

I'm probably reinventing the wheel here, so I'll stop babbling...

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 2 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2159]: Sorry, GitHub PR view tricked me into thinking this was some other operator. Yeah, that makes sense.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala (hunk 1 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala:40]: Why do we need `andThen` for that?

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala:40]: When use `recoverWith` will be :
```scala
(pf:[Throwable, Boolean]).andThen({
case true => Source.empty
}
```

**Comment by He-Pin** [comment]
@patriknw ping :)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 2 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:2159]: As this PR may need more time, I will extract this change to a separate pr.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 3]
```diff
@@ -1803,6 +1803,41 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
...
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with predicate return ture
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(predicate: java.util.function.Predicate[_ >: Throwable]): javadsl.Flow[In, Out, Mat] =
+    new Flow(delegate.recoverWithRetries(-1, {
+      case ex: Throwable if predicate.test(ex) => scaladsl.Source.empty
+    }))
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:1839]: Define a name attribute for it and set on the composed stage, also, define in one place (FlowOps) and delegate to that rather than repeat impl in Java DSL etc.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 4]
```diff
@@ -1803,6 +1803,41 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
...
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with predicate return ture
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(predicate: java.util.function.Predicate[_ >: Throwable]): javadsl.Flow[In, Out, Mat] =
+    new Flow(delegate.completeOn {
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:1837]: This will not work on 2.12.x

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 5]
```diff
@@ -882,6 +882,46 @@ trait FlowOps[+Out, +Mat] {
...
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or fails with exception is an instance of the provided type
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn[T <: Throwable](implicit tag: ClassTag[T]): Repr[Out] = completeOn {
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:899]: should I change it to 
```scala
def completeOn()[T <: Throwable](implicit tag: ClassTag[T])
```
otherwise not compile on 2.12.x

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:899]: I guess that might be fine, `.completeOn[SomeEx]()` doesn't look too bad IMO

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 6]
```diff
@@ -882,6 +882,46 @@ trait FlowOps[+Out, +Mat] {
...
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(pf: PartialFunction[Throwable, Boolean]): Repr[Out] =
+    via(
+      Flow[Out]
+        .recoverWithRetries(-1, pf.andThen((_: Boolean @unchecked) match {
+          case true => Source.empty
+        }))
+        .withAttributes(DefaultAttributes.completeOn and SourceLocation.forLambda(pf)))
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:923]: Or we give this one a separate name

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 5 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:899]: Or rewrite it as 
```
    val pf: PartialFunction[Throwable, Boolean] = {
      case ex if tag.runtimeClass.isInstance(ex) => true
    }
    completeOn(pf)
```
?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 6 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:925]: `completeWhen`?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 5 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Yeah, https://github.com/scala/bug/issues/12688, seems the only way to go now is  :
```
def completeOn()[T <: Throwable](implicit tag: ClassTag[T])
```

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 6 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:925]: @johanandren How about `completeWhen`?

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:923]: `completeWhen`?

**Review by johanandren — APPROVED** [review]
LGTM, lets go with the empty param list like this 👍

**Comment by He-Pin** [comment]
@patriknw ping~

**Comment by He-Pin** [comment]
@patriknw Any input on this?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 6 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:923]: I think that's a good name

**Comment by He-Pin** [comment]
@patriknw I have rebased on to the current main

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala (hunk 6 above)
Comment (octonato) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:921]: Are we still planning to name it `completeWhen` for the PF variation?

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:921]: @octonato I think so.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:921]: @johanandren But that way, the scaladsl and javadsl will have different name.

Comment (octonato) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:921]: The predicate variant in the javadsl can also be `completeWhen`. Why not?

I think it makes lots of sense that a predicated one is suffixed with `When`.

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala [hunk 7]
```diff
@@ -2292,6 +2292,41 @@ final class Source[Out, Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[
...
+   *
+   * '''Emits when''' element is available from the upstream
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or failed with predicate return ture
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def completeOn(predicate: java.util.function.Predicate[_ >: Throwable]): javadsl.Source[Out, Mat] =
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: I know the naming has been discussed, but this name doesn't include much about that it is about errors. Might be difficult to discover. `onErrorComplete` is better. I still like `recoverWithComplete` because it's close to what we already have, but seems like @johanandren didn't like it.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: There are :
- `onErrorComplete(..)`
- `onErrorContinue(...)`
- `onErrorMap(...)`
- `onErrorReturn(...)`
- `onErrorResume(...)`
- `onErrorStop(...)`

in reactor-core.  and should we use `completeOn` , `completeWhen`, or `recoverWithComplete` here, naming is always hard.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: Both `CompleteOn` and `CompleteWhen` seems reasonable to accept a predicate/condition, but for `recoverWithComplete` for me I think it should take zero parameters.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: @octonato What's your input? the `recoverWithComplete` align better but seems like `completeWhen` is a great name too.

Comment (octonato) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: I agree that `onErrorComplete` is better for discoverability and it express much better that the method has the intention to be used in case of an error. 

About `recoverWithComplete` is less interesting, I think. Internally, we are 'recovering' with empty Source, but that's an implementation detail. 

From a user perspective, I don't think recover makes sense because we are not recovering (and continuing) when an error occurs. We are completing gracefully. 

That said, maybe a overloaded `onErrorComplete` (Predicated and PF) will do the job. Clear intention and discoverable.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: @johanandren I am going to update this PR with `onErrorComplete` , as the https://github.com/akka/akka-persistence-r2dbc is using `Flux` and in `Flux` it's named `onErrorComplete` too.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:2325]: I asked chatgpt, it suggests `completeWhen`.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: @johanandren I updated the PR to `onErrorComplete` and the https://github.com/akka/akka-persistence-r2dbc is using `Flux` too.

And as we here we only matching on error types, and `completeOn`/`completeWhen` is more generic.

**Review by octonato — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala (hunk 7 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: Let's go with onErrorComplete

**Review by johanandren — APPROVED** [review]
